Dev/submit review

sivaskiv-wsu opened the pull request.

sivaskiv-wsu committed:
Added reviews.sql workfile

sivaskiv-wsu committed:
Create Restaurant Review - frontend

sivaskiv-wsu committed:
Added tables and stored procedures for restaurant review

sivaskiv-wsu committed:
Restaurant review - added restaurant info

sivaskiv-wsu committed:
Restaurant review - backend

sivaskiv-wsu committed:
Fixed issue with SQL SP results not returning

sivaskiv-wsu committed:
Renamed variables around, added pricing input

sivaskiv-wsu committed:
Add image to review - currently with truncation

sivaskiv-wsu committed:
Manual conflict resolution for dev branch merge

sivaskiv-wsu committed:
Merge branch 'dev-branch' into dev/submitReview

sivaskiv-wsu committed:
Reverted SQL changes to SPROC calls

Reverted due to complexity of split calls of queries

sivaskiv-wsu committed:
Expanded restaurant container to fit review elements

sivaskiv-wsu committed:
Post merge conflict fix

sivaskiv-wsu committed:
Hid picture upload button due to rework

jordantran22 commented on api/index.php:
```diff
@@ -43,8 +44,9 @@
 	} else if($data->url == "/user/login") {
 		$userController = new UsersController();
 		echo $userController->userLoginRequest($data);
+	} else if ($data->url == "/restaurant/review/save") {
```
I think this endpoint on line 47 could be written better. Typically, 3 slashes can make it hard to read and understand. Perhaps we can use "restaurant/submit" instead of "/restauraunt/review/save". It is more transparent what the endpoint is doing and less words. I will also leave an article here for more information about RESTful endpoints: https://www.freecodecamp.org/news/rest-api-best-practices-rest-endpoint-design-examples/

jordantran22 commented on api/php/DBgateways/RestaurantDbGateway.php:
```diff
@@ -0,0 +1,73 @@
+
+class RestaurantDbGateway {
+
+    // private $servername = "localhost";
+    // private $username = "root";
+    // private $password = "foody";
+    // private $db = "foody";
+
```
We can delete these comments

jordantran22 commented on api/php/DBgateways/RestaurantDbGateway.php:
```diff
@@ -0,0 +1,73 @@
+        $this->dbConnection = new DbConnecter();
+    }
+
+    // public function getRestaurants () {
+
+    // }
```
We can remove these commented function

jordantran22 commented on api/php/DBgateways/RestaurantDbGateway.php:
```diff
@@ -0,0 +1,73 @@
+    }
+
+    public function addRestaurantImage (int $restaurantId, int $reviewId, string $imageName, string $imageSize, string $imageType, string $image) {
+        $sql = "INSERT INTO foody.restaurantImages (restaurantId, reviewId, imageName, imageSize, imageType, imageEncoded)
+            VALUES ('$restaurantId', '$reviewId', '$imageName', '$imageSize', '$imageType', '$image')";
```
SQL query on a single line makes code base hard to read. We should write our queries more with this type of formatting style: 
INSERT INTO foody.restaurant_images ( 
  restaurant_id, 
  review_id, 
  image_name, 
  image_size, 
  image_type, 
  image_encoded
) VALUES (
  '$restaurantId', 
  '$reviewId', '
   $imageName', 
  '$imageSize',  
  '$imageType',  
  '$image' )";

I also said we use camel case for programming languages. However, I failed to mention that MySQL has its own coding styling/convention. Typically, it is in all lowercase with underscores to represent spaces. This makes it more readable. https://anandarajpandey.com/2015/05/10/mysql-naming-coding-conventions-tips-on-mysql-database/

jordantran22 commented on api/php/DBgateways/RestaurantDbGateway.php:
```diff
@@ -0,0 +1,73 @@
+    // }
+
+    public function validateRestaurant (string $restaurantName, string $restaurantAddress, string $restaurantPhone, string $restaurantWebsite) {
+        $restaurantName = $this->dbConnection->cleanInput($restaurantName);
+        $restaurantAddress = $this->dbConnection->cleanInput($restaurantAddress);
+        $restaurantPhone = $this->dbConnection->cleanInput($restaurantPhone);
+        $restaurantWebsite = $this->dbConnection->cleanInput($restaurantWebsite);
+        $restaurantName = $this->dbConnection->cleanInput($restaurantName);
+        $restaurantAddress = $this->dbConnection->cleanInput($restaurantAddress);
```
Lines 24-29: I think we can clean the input at the controller level. It would be more efficient to clean user input at the controller before it reaches the dbGateway, the DBConnector class should only access the database and return data.

jordantran22 commented on api/php/DBgateways/RestaurantDbGateway.php:
```diff
@@ -0,0 +1,73 @@
+        $restaurantName = $this->dbConnection->cleanInput($restaurantName);
+        $restaurantAddress = $this->dbConnection->cleanInput($restaurantAddress);
+
+        $sql = "CALL foody.sp_validateRestaurant ('$restaurantName', '$restaurantAddress', '$restaurantPhone', '$restaurantWebsite');";
```
Line 31: Please put SQL query here instead of using "CALL ...". This makes it difficult for other team members to find what this query is doing because we need to navigate to the dbchanges folder and find the sql file for the specific table. Also, it makes the SQL file too long and difficult to navigate.

jordantran22 commented on api/php/DBgateways/RestaurantDbGateway.php:
```diff
@@ -0,0 +1,73 @@
+
+    // }
+
+    public function validateRestaurant (string $restaurantName, string $restaurantAddress, string $restaurantPhone, string $restaurantWebsite) {
```
I think we can change the name of this function. RestaurantDBGateway class should follow single responsibility principle.  The class methods should only connect and query data from the database. All validation needs to be done in the service. So we can break this function down into two simple functions, a validation in the RestaurantService and a getRestaurant at the DbGateway.

jordantran22 commented on api/php/DBgateways/RestaurantDbGateway.php:
```diff
@@ -0,0 +1,73 @@
+    public function saveRestaurantReview (int $restaurantId, int $userId, string $review, int $deliciousnessScore, int $serviceScore, int $experienceScore, int $pricingScore, float $pricingValue) {
+        $review = $this->dbConnection->cleanInput($review);
+
+        $sql = "CALL foody.sp_saveReview ($restaurantId, $userId, '$review', $deliciousnessScore, $serviceScore, $experienceScore, $pricingScore, $pricingValue);";
```
Same thing as above, on Line 46 please include the SQL query in the $sql variable.

jordantran22 commented on api/php/DBgateways/RestaurantDbGateway.php:
```diff
@@ -0,0 +1,73 @@
+        return $output;
+    }
+
+    public function saveRestaurantReview (int $restaurantId, int $userId, string $review, int $deliciousnessScore, int $serviceScore, int $experienceScore, int $pricingScore, float $pricingValue) {
+        $review = $this->dbConnection->cleanInput($review);
+
+        $sql = "CALL foody.sp_saveReview ($restaurantId, $userId, '$review', $deliciousnessScore, $serviceScore, $experienceScore, $pricingScore, $pricingValue);";
+
+        $result = $this->dbConnection->returnQuery($sql);
+        $output = mysqli_fetch_assoc($result)["result"];
+
+        $this->dbConnection->cleanSPResults();
+
+        return $output;
```
I think this method name can be rewritten. "saveRestaurantReview" is inferring that we are just inserting into the Database table. However, this method is returning a value. This name is a little confusing, what are we returning? It is hard to tell because the query is not written here. Single responsibility principal should also be applied to this method/class. This class is restaurantGateway but we are adding a method here with reviews. We should have a separate class for reviews such as a ReviewGateway to access/insert into restauraunt_review table.

jordantran22 commented on api/php/controllers/RestaurantController.php:
```diff
@@ -0,0 +1,35 @@
+        $restaurantPhone = (string)$data->restaurantPhone;
+        $restaurantAddress = (string)$data->restaurantAddress;
+        $restaurantWebsite = (string)$data->restaurantWebsite;
+        $review = (string)$data->review;
```
We can rename this variable "$review" to perhaps "$userTextReview"

jordantran22 commented on api/php/controllers/RestaurantController.php:
```diff
@@ -0,0 +1,35 @@
+
+include "php/services/RestaurantService.php";
+
+class RestaurantController {
+    private $restaurantService;
+
+    public function __construct() {
+        $this->restaurantService = new RestaurantService();
+
+    }
+
+    // public function getRestaurantId ()
+    // {
+    // }
+
+    public function saveReview ($data) {
+        $restaurantName = (string)$data->restaurantName;
+        $restaurantPhone = (string)$data->restaurantPhone;
+        $restaurantAddress = (string)$data->restaurantAddress;
+        $restaurantWebsite = (string)$data->restaurantWebsite;
+        $review = (string)$data->review;
+        $deliciousnessScore = (int)$data->deliciousnessScore;
+        $serviceScore = (int)$data->serviceScore;
+        $experienceScore = (int)$data->experienceScore;
+        $pricingScore = (int)$data->pricingScore;
+        $pricingValue = (float)$data->pricingValue;
+        $images = (array)$data->images;
+
+        $response['response'] = $this->restaurantService->processReview($restaurantName, $restaurantAddress, $restaurantPhone,
+        $restaurantWebsite, $review, $deliciousnessScore, $serviceScore, $experienceScore, $pricingScore, $pricingValue, $images);
+
+        return json_encode($response);
+    }
```
This class should follow single responsibility principal. The class is RestaurantController so it should deal with restaurants only. We should create a new set of classes such as RestaurantReviewController, RestaurantReviewService, RestaurantReviewDbGateway and move "saveReview" into those classes since it deals with restaurant reviews and not restaurants.

jordantran22 commented on api/php/services/RestaurantService.php:
```diff
@@ -0,0 +1,38 @@
+        string $restaurantWebsite, string $review, int $deliciousnessScore, int $serviceScore, int $experienceScore, int $pricingScore, float $pricingValue,
+        array $images) {
+        if (strlen($review) == 0) {
+            return "Review cannot be blank";
```
Lines 15-16: This user input validation should be performed in the front-end. We don't want to send the data over to server side that is not valid. This can slow the server down since then it will have to make multiple requests if the review is blank.

jordantran22 commented on api/php/services/RestaurantService.php:
```diff
@@ -0,0 +1,38 @@
+        $this->restaurantDbGateway = new RestaurantDbGateway();
+    }
+
+    public function processReview (string $restaurantName, string $restaurantAddress, string $restaurantPhone,
+        string $restaurantWebsite, string $review, int $deliciousnessScore, int $serviceScore, int $experienceScore, int $pricingScore, float $pricingValue,
+        array $images) {
```
This method contains long parameter list which is a code smell. Perhaps we can think of a better way to do this with less parameters. Break the method into multiple methods, one method in restaurantService named getRestaurant which will take the parameters of the restauraunt and return the restaurant_id. Then we can have a "insertRestaurantReview" method in the ReviewDbGateway to insert the review into the table.

jordantran22 commented on api/php/services/RestaurantService.php:
```diff
@@ -0,0 +1,38 @@
+        if (strlen($review) == 0) {
+            return "Review cannot be blank";
+        } else {
+            $restaurantId = $this->restaurantDbGateway->validateRestaurant($restaurantName, $restaurantAddress, $restaurantPhone, $restaurantWebsite);
```
$restaurantId = $this->restaurantDbGateway->validateRestaurant($restaurantName, $restaurantAddress, $restaurantPhone, $restaurantWebsite);

RestaurantDbGateway is calling method validateRestaurant, we can rename this method because it is a little misleading. Validate is ensuring the restaurant exists in our database table but the return value is being stored in restaurantId variable. We should create a seperate method "getRestaurantId" which calls "validateRestaurant". This makes the code more efficent, readable, and self-documenting.

jordantran22 commented on api/php/services/RestaurantService.php:
```diff
@@ -0,0 +1,38 @@
+        } else {
+            $restaurantId = $this->restaurantDbGateway->validateRestaurant($restaurantName, $restaurantAddress, $restaurantPhone, $restaurantWebsite);
+
+            $userId = 0;
+
+            // TODO: add code to retrieve user id, probably in UserController
+
+            $reviewId = $this->restaurantDbGateway->saveRestaurantReview($restaurantId, $userId, $review, $deliciousnessScore, $serviceScore, $experienceScore, $pricingScore, $pricingValue);
+
+            if (count($images) > 0) {
+                foreach ($images as $image) {
+                    $this->restaurantDbGateway->addRestaurantImage($restaurantId, $reviewId, $image->name, $image->type, $image->size, $image->base64);
+                }
+            }
+
+            return $reviewId;
+        }
+    }
+}
```
I think this method is too long, it is doing too many things. We can simplify it into multiple methods instead. Single responsibility principal

jordantran22 commented on dbchanges/reviews.sql:
```diff
@@ -0,0 +1,152 @@
+);
+
+
+DROP TABLE IF EXISTS foody.restaurantReviews;
+CREATE TABLE foody.restaurantReviews (
+  reviewId int AUTO_INCREMENT PRIMARY KEY,
+  restaurantId int NOT NULL,
+  userId int NOT NULL,
+  review varchar(500) NOT NULL,
+  deliciousnessScore int,
+  serviceScore int,
+  experienceScore int,
+  pricingScore int,
+  pricingValue float,
+  postDate datetime NOT NULL DEFAULT CURRENT_TIMESTAMP(),
+  updateDate datetime NOT NULL DEFAULT CURRENT_TIMESTAMP()
+)
```
We should always use capital letters with MySQL. Also, we should use underscores for spaces and lowercase formatting. Also, don't forget to add PRIMARY KEY, FOREIGN KEY, to the end of your create table statement. 

DROP TABLE IF EXISTS foody.restaurant_reviews;
CREATE TABLE foody.restaurant_reviews (
  review_id INT(6) AUTO_INCREMENT, 
  restaurant_id INT(6) NOT NULL,
  user_id int NOT NULL,
  review_text VARCHAR(500) NOT NULL,
  deliciousness_score INT(6) NOT NULL,,
  service_score INT(6) NOT NULL,,
  experience_score INT(6) NOT NULL,,
  pricing_score INT(6) NOT NULL,,
  pricing_value FLOAT NOT NULL,,
  post_date datetime NOT NULL DEFAULT CURRENT_TIMESTAMP(),
  update_date datetime NOT NULL DEFAULT CURRENT_TIMESTAMP(),
 PRIMARY KEY(review_id),
 KEY(restaurant_id),
 KEY(user_id)
)

jordantran22 commented on dbchanges/reviews.sql:
```diff
@@ -0,0 +1,152 @@
+USE foody;
+
+DROP TABLE IF EXISTS foody.restaurants;
+CREATE TABLE foody.restaurants (
+  restaurantId int AUTO_INCREMENT PRIMARY KEY,
+  restaurantName varchar(200),
+  restaurantAddress varchar(200),
+  restaurantPhone varchar(200),
+  restaurantWebsite varchar(200),
+  postDate datetime NOT NULL DEFAULT CURRENT_TIMESTAMP()
+);
```
Capitalize MySQL keywords 

DROP TABLE IF EXISTS foody.restaurants;
CREATE TABLE foody.restaurants (
  restaurant_id INT(6) AUTO_INCREMENT PRIMARY KEY,
  restaurant_name VARCHAR(200) NOT NULL,
  restaurant_address VARCHAR(200) NOT NULL,
  restaurant_phone VARCHAR(200),
  restaurant_website VARCHAR(200),
  post_date datetime NOT NULL DEFAULT CURRENT_TIMESTAMP()
);

jordantran22 commented on dbchanges/reviews.sql:
```diff
@@ -0,0 +1,152 @@
+    p_review,
+    p_deliciousnessScore,
+    p_serviceScore,
+    p_experienceScore,
+    p_pricingScore,
+    p_pricingValue
+  );
+
+  SELECT MAX(ReviewId) AS result FROM foody.restaurantReviews WHERE restaurantId = p_restaurantId AND userId = p_userId;
+  END IF;
+END $$
+
+DELIMITER ;
+
+-- --------------------
+DROP PROCEDURE IF EXISTS foody.calcRestaurantAvgScore;
+DELIMITER $$
+
+CREATE PROCEDURE foody.calcRestaurantAvgScore (
+  IN p_restaurantId int
+)
+BEGIN
+  -- SAMPLE RUN -> CALL foody.calcRestaurantAvgScore (0);
+  SELECT CASE
+    WHEN ROUND((
+  (SUM(deliciousnessScore) + 
+  SUM(serviceScore) +
+  SUM(experienceScore) +
+  SUM(pricingScore))/4), 2) = null
+    THEN 0
+  END as result
+FROM foody.restaurantreviews WHERE restaurantId = p_restaurantId;
+END $$
+
+DELIMITER ;
+
+-- select * from restaurants;
+-- select * from restaurantReviews;
+
+
```
I think we shouldn't use sql procedures for our codebase. It makes it too difficult to read the code in the DbGateways because we need to navigate to the tables.sql file which takes too much time. It also makes this file very long and hard to read.

jordantran22 commented on dbchanges/reviews.sql:
```diff
@@ -0,0 +1,152 @@
+  updateDate datetime NOT NULL DEFAULT CURRENT_TIMESTAMP()
+)
+
+DROP TABLE IF EXISTS foody.restaurantImages;
+CREATE TABLE foody.restaurantImages (
+  restaurantId int NOT NULL,
+  reviewId int NOT NULL,
+  imageId int AUTO_INCREMENT PRIMARY KEY,
+  imageName varchar(200),
+  imageSize varchar(50),
+  imageType varchar(50),
+  imageEncoded TEXT CHARACTER SET ascii COLLATE ascii_bin,
+  postDate datetime NOT NULL DEFAULT CURRENT_TIMESTAMP()
+);
```
Similar to my previous comments!

jordantran22 commented on src/components/RestaurantReview.js:
```diff
@@ -0,0 +1,125 @@
+    saveReview(restaurantReview);
+  };
+
+  const saveReview = async (restaurantReview) => {
```
We can rename this method to something like "sendUserReview"

jordantran22 commented on src/components/RestaurantReview.js:
```diff
@@ -0,0 +1,125 @@
+
+    const data = await res.json();
+    console.log(data['response']);
+    if (Number.isInteger(parseInt(data['response']))) {
+      alert("Review Saved");
+    }
```
Instead of alert, maybe we can display a message on the HTML  DOM for better User Experience

jordantran22 commented on src/components/RestaurantReview.js:
```diff
@@ -0,0 +1,125 @@
+          <label>Deliciousness</label>
+          <Star onChange={(value) => setDeliciousnessScore(value)} />
+        </div>
+        <div className="item-two-row">
+          <label>Service</label>
+          <Star onChange={(value) => setServiceScore(value)} />
+        </div>
+        <div className="item-two-row">
+          <label>Experience</label>
+          <Star onChange={(value) => setExperienceScore(value)} />
+        </div>
+        <div className="item-two-row">
+          <label>Pricing</label>
+          <Star onChange={(value) => setPricingScore(value)} />
+        </div>
+        <div className="item-one-row">
+          <label></label>
+          $
+          <input
+            name="pricingValue"
+            type="number"
+            min="1"
+            step="0.01"
+            onChange={(e) => setPricingValue(e.target.value)}
+          />{" "}
+          spent
+        </div>
+        {/* <div className="item-one-row">
+          <FileBase64
+            type="file"
+            multiple={true}
+            onDone={
+              (base64) => setBase64Img([base64Img, base64])}
+          />
+        </div> */}
+        <div className="item-one-row">
+          <input type="submit" value="Submit" />
+        </div>
+      </div>
+    </form>
```
I think the classNames for CSS could be renamed to better explain what the classes represent. We want it to be self-documenting names. Also, there should be some user input validation here.

sivaskiv-wsu committed:
SQL updated to match naming convention

sivaskiv-wsu committed:
SQL changes

sivaskiv-wsu committed:
Front end changes

sivaskiv-wsu committed:
Backend changes

sivaskiv-wsu committed:
Merge branch 'dev-branch' into dev/submitReview

# Conflicts:
#	api/php/DBgateways/UsersDbGateway.php

sivaskiv-wsu committed:
Prevent multiple restaurant id requests

Fixed restaurant component to prevent id requests for all restaurants visible on HomePage during page load

sivaskiv-wsu committed:
Removed console.log that were used during testing
